crates/bevy_ecs/src/schedule/system_graph.rs

```diff
@@ -0,0 +1,333 @@
+        self.create_node(system.into())
+    }
+
+    /// Checks if two graphs instances point to the same logical underlying graph.
+    pub fn is_same_graph(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+
+    fn create_node(&self, mut system: SystemDescriptor) -> SystemGraphNode {
+        let mut nodes = self.nodes.borrow_mut();
+        assert!(
+            nodes.len() <= u32::MAX as usize,
+            "Cannot add more than {} systems to a SystemGraph",
+            u32::MAX
+        );
+        let id = NodeId {
+            graph_id: self.id,
+            node_id: nodes.len() as u32,
+        };
+        match &mut system {
+            SystemDescriptor::Parallel(descriptor) => descriptor.labels.push(id.dyn_clone()),
+            SystemDescriptor::Exclusive(descriptor) => descriptor.labels.push(id.dyn_clone()),
+        }
+        nodes.insert(id, system);
+        SystemGraphNode {
+            id,
+            graph: self.clone(),
+        }
+    }
+
+    fn add_dependency(&self, src: NodeId, dst: NodeId) {
+        if let Some(system) = self.nodes.borrow_mut().get_mut(&dst) {
+            match system {
+                SystemDescriptor::Parallel(descriptor) => descriptor.after.push(src.dyn_clone()),
+                SystemDescriptor::Exclusive(descriptor) => descriptor.after.push(src.dyn_clone()),
+            }
+        } else {
+            panic!(
+                "Attempted to add dependency for {:?}, which doesn't exist.",
+                dst
+            );
+        }
+    }
+}
+
+/// A draining conversion to [SystemSet]. All other clones of the same graph will be empty
+/// afterwards.
+///
+/// [SystemSet]: crate::schedule::SystemSet
+impl From<SystemGraph> for SystemSet {
+    fn from(graph: SystemGraph) -> Self {
+        let mut system_set = SystemSet::new();
+        for (_, system) in graph.nodes.borrow_mut().drain() {
+            system_set = system_set.with_system(system);
+        }
+        system_set
+    }
+}
+
+#[derive(Clone)]
+pub struct SystemGraphNode {
+    id: NodeId,
+    graph: SystemGraph,
+}
+
+impl SystemGraphNode {
+    /// Creates a new node in the graph and adds the current node as it's dependency.
+    ///
+    /// This function can be called multiple times to add mulitple systems to the graph,
+    /// all of which will not execute until original node's system has finished running.
+    pub fn then(&self, next: impl Into<SystemDescriptor>) -> SystemGraphNode {
+        let node = self.graph.create_node(next.into());
+        self.graph.add_dependency(self.id, node.id);
+        node
+    }
+
+    /// Fans out from the given node into multiple dependent systems. All provided
+    /// systems will not run until the original node's system finishes running.
+    ///
+    /// Functionally equivalent to calling `SystemGraphNode::then` multiple times.
+    pub fn split_into<T: SystemGroup>(&self, system_group: T) -> T::Output {
+        system_group.split_into(self)
+    }
```

Comment:
Nice work on getting the tuple api to work :)
However, why do we need to split the `then`/`split_into` APIs?

Ideally I'd want something like:
```rust
let root = SystemGraph::new().root(a.system()).then(b.system()).then(c.system());
```
this constructs a graph:
```
A -> B -> C
```

however I would also like to do:
```rust
let root = SystemGraph::new()
    .root(a.system())
    .then((b1.system(), b2.system()))
    .then(c.system());
```
which would construct a 
```
   / B1  \
A          C
   \ B2  /
```
e.g. the type that is returned from `then` should be able to either chain a `then` command where the previously added system(s) (`B1` and `B2`) are run in parallel and then the newly added `C` system is run after `B1` and `B2`

Optionally, then we should be able to break apart `B1` and `B2` and add independent chains e.g.
```rust
let (b1, b2) = SystemGraph::new()
    .root(a.system())
    .then((b1.system(), b2.system()));

b1.then(c1.system());
b2.then(c2.system());
```
which would produce
```
   / B1  -> C1
A          
   \ B2  -> C2
```

Perhaps this might be over engineering, but I like keeping the interface similar. 
Though optionally, perhps `split_into` should be named `fork` or something like that?

Reply:
I intentionally chose this to avoid potentially surprising behavior like the following:

```
let (b1, b2) = SystemGraph::new() 
  .root(a.system()) 
  .then((b1.system(), b2.system()))
  .then((c1.system(), c2.system()));
```

where it's unclear if it would produce something like what you have in the second example, or the following:

```
    / B1 -> C1
A          X
    \ B2 -> C2
```

I definitely like the matchup of `fork`/`join` though. I'll go ahead and rename the function.

Reply:
For the one -> many -> one case, I added implementations for joining the output of `fork` together. This is now valid:

```rust
let root = SystemGraph::new()
    .root(a.system())
    .fork((b1.system(), b2.system()))
    .join(c.system());
```

This should allow us to make any one to one, one to many, and many to one dependencies, but any many to many dependencies need to manually constructed.